### ISSUE TITLE
AMQP API blocks while network failure

### ISSUE THREAD

**Issue by rohitkp** [alanxz/rabbitmq-c#175]
While sending big file around 6 MB. If the network error like eth0 cable pulled out or network router is switched off the API `amqp_simple_publish()` will hang for infinite time. Likewise while receiving the same file in third while loop i.e (body_received < body_target) the AMQP API `amqp_simple_wait_frame_noblock(conn, &frame, &t)` gets block for infinite time.

Please raised it as Bug.

 for your reference here is the publish and listen programs.

```c
/* AMQPS SEND PROGRAM to simulate Publish hang while publishing big file */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <stdint.h>
#include <amqp_ssl_socket.h>
#include <amqp_framing.h>

#include "utils.h"

int main(int argc, char const *const *argv)
{
  char const *hostname;
  int port, status,ret;
  char const *exchange;
  char const *routingkey;
  //char const *messagebody;
  amqp_socket_t *socket;
  amqp_connection_state_t conn;


  FILE *fp = NULL;
  struct stat File_Stat;
  char file_name[250];
  amqp_bytes_t msg;
  char *buffer;

  if (argc < 5) {
    fprintf(stderr, "Usage: amqps_sendstring host port exchange routingkey "
            "messagebody [cacert.pem [key.pem cert.pem]]\n");
    return 1;
  }

  hostname = argv[1];
  port = atoi(argv[2]);
  exchange = argv[3];
  routingkey = argv[4];
  //messagebody = argv[5];

  conn = amqp_new_connection();

  socket = amqp_ssl_socket_new(conn);

  socket = amqp_ssl_socket_new(conn);
  
  if (!socket) {
    die("creating SSL/TLS socket");
  }

  status = amqp_ssl_socket_set_cacert(socket, argv[5]);
  if (status) {
 	     die("setting CA certificate");
   	}
 
  status = amqp_ssl_socket_set_key(socket,argv[7], argv[6]);
  if (status) {
	      die("setting client cert");
    	}
  
  amqp_ssl_socket_set_verify(socket, 0);
  status = amqp_socket_open(socket, hostname, port);
  if (status) {
    		die("opening SSL/TLS connection");
  	}

  amqp_ssl_socket_set_verify(socket, 0);
  amqp_login(conn, "/", 0, 131072, 0, AMQP_SASL_METHOD_PLAIN, "guest", "guest");
                    
  amqp_channel_open(conn, 1);
  die_on_amqp_error(amqp_get_rpc_reply(conn), "Opening channel");

    printf("Enter the name of file you wish to send: ");
    gets(file_name);

    fp = fopen(file_name,"r"); // read mode
    if( fp == NULL )
    {
	  printf("Cannot open file\n");
	  return 0;
    }

    stat(file_name,&File_Stat);
    printf("File Size is %d\n",File_Stat.st_size);
    buffer = malloc(File_Stat.st_size);
    if (buffer == NULL) {
      	  printf("Error in malloc\n");
      	  exit (2);
     }

    static int count = 0;
    amqp_basic_properties_t props;
    props._flags = AMQP_BASIC_CONTENT_TYPE_FLAG | AMQP_BASIC_DELIVERY_MODE_FLAG;
    props.content_type = amqp_cstring_bytes("text/plain");
    props.delivery_mode = 2; /* persistent delivery mode */

    
     size_t result = fread (buffer,1,File_Stat.st_size,fp);
     msg.len = File_Stat.st_size;
     printf("Data len : %d\n", msg.len);
     msg.bytes=buffer;
//     printf("msg.bytes:%s\n",buffer);

    printf("Publishing...\n");
    ret=amqp_basic_publish(conn,
                                    1,
                                    amqp_cstring_bytes(exchange),
                                    amqp_cstring_bytes(routingkey),
                                    1,
                                    0,
                                    &props,
                                    msg);
               
    count++;
    printf("Publish count %d\n",count);	

  die_on_amqp_error(amqp_channel_close(conn, 1, AMQP_REPLY_SUCCESS), "Closing channel");
  die_on_amqp_error(amqp_connection_close(conn, AMQP_REPLY_SUCCESS), "Closing connection");
  die_on_error(amqp_destroy_connection(conn), "Ending connection");
  return 0;
}
```

```c
/* AMQPS LISTEN PROGRAM to simulate amqp_simple_wait_frame_noblock() hang while listening big data */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <amqp_ssl_socket.h>
#include <amqp_framing.h>
#include <assert.h>
#include "utils.h"

int main(int argc, char const *const *argv)
{
  char const *hostname;
  int port, status;
  char const *exchange;
  char const *bindingkey;
  amqp_socket_t *socket;
  amqp_connection_state_t conn;

  amqp_bytes_t queuename;

  if (argc < 5) {
    fprintf(stderr, "Usage: amqps_listen host port exchange bindingkey "
            "[cacert.pem [key.pem cert.pem]]\n");
    return 1;
  }

  hostname = argv[1];
  port = atoi(argv[2]);
  exchange = argv[3];
  bindingkey = argv[4];

  conn = amqp_new_connection();

  socket = amqp_ssl_socket_new(conn);
  if (!socket) {
    die("creating SSL/TLS socket");
  }

  if (argc > 5) {
    status = amqp_ssl_socket_set_cacert(socket, argv[5]);
    if (status) {
      die("setting CA certificate");
    }
  }

  if (argc > 7) {
    status = amqp_ssl_socket_set_key(socket, argv[7], argv[6]);
    if (status) {
      die("setting client cert");
    }
  }

  amqp_ssl_socket_set_verify(socket, 0);
  printf("HostName:%s ,POrtNo:%d\n",hostname,port);
  status = amqp_socket_open(socket, hostname, port);
  if (status) {
    die("opening SSL/TLS connection");
  }

  die_on_amqp_error(amqp_login(conn, "/", 0, 131072, 0, AMQP_SASL_METHOD_PLAIN, "guest", "guest"),
                    "Logging in");
  amqp_channel_open(conn, 1);
  die_on_amqp_error(amqp_get_rpc_reply(conn), "Opening channel");

  {
    amqp_queue_declare_ok_t *r = amqp_queue_declare(conn, 1, amqp_empty_bytes, 0, 0, 0, 1,
                                 amqp_empty_table);
    die_on_amqp_error(amqp_get_rpc_reply(conn), "Declaring queue");
    queuename = amqp_bytes_malloc_dup(r->queue);
    if (queuename.bytes == NULL) {
      fprintf(stderr, "Out of memory while copying queue name");
      return 1;
    }
  }

  amqp_queue_bind(conn, 1, queuename, amqp_cstring_bytes(exchange), amqp_cstring_bytes(bindingkey),
                  amqp_empty_table);
  die_on_amqp_error(amqp_get_rpc_reply(conn), "Binding queue");

  amqp_basic_consume(conn, 1, queuename, amqp_empty_bytes, 0, 1, 0, amqp_empty_table);
  die_on_amqp_error(amqp_get_rpc_reply(conn), "Consuming");

  {
    amqp_frame_t frame;
    int result;

    amqp_basic_deliver_t *d;
    amqp_basic_properties_t *p;
    size_t body_target;
    size_t body_received;
	struct timeval t;
	t.tv_sec=2;
	t.tv_usec=10;
	int count=0;

    while (1) {
      amqp_maybe_release_buffers(conn);
	  result = amqp_simple_wait_frame_noblock(conn, &frame, &t);
      
      printf("Result %d\n", result);
      if (result < 0) {
       		continue;// break;
      }

      printf("Frame type %d, channel %d\n", frame.frame_type, frame.channel);
      if (frame.frame_type != AMQP_FRAME_METHOD) {
        continue;
      }

      printf("Method %s\n", amqp_method_name(frame.payload.method.id));
      if (frame.payload.method.id != AMQP_BASIC_DELIVER_METHOD) {
        continue;
      }

      d = (amqp_basic_deliver_t *) frame.payload.method.decoded;
      printf("Delivery %u, exchange %.*s routingkey %.*s\n",
             (unsigned) d->delivery_tag,
             (int) d->exchange.len, (char *) d->exchange.bytes,
             (int) d->routing_key.len, (char *) d->routing_key.bytes);

	result = amqp_simple_wait_frame_noblock(conn, &frame, &t);      

      if (result < 0) {
        break;
      }

      if (frame.frame_type != AMQP_FRAME_HEADER) {
        fprintf(stderr, "Expected header!");
        abort();
      }
      p = (amqp_basic_properties_t *) frame.payload.properties.decoded;
      if (p->_flags & AMQP_BASIC_CONTENT_TYPE_FLAG) {
        printf("Content-type: %.*s\n",
               (int) p->content_type.len, (char *) p->content_type.bytes);
      }
      printf("----\n");

      body_target = frame.payload.properties.body_size;
      body_received = 0;
      count=0;
      printf("body_target = %d\n", body_target);
      printf("Number of frames expected = %d\n", (body_target/131072));

      while (body_received < body_target) {
    	  printf("Count = %d body_received = %d\n", count, body_received);
		result = amqp_simple_wait_frame_noblock(conn, &frame, &t);

        if (result < 0) {
          break;
        }

        if (frame.frame_type != AMQP_FRAME_BODY) {
          fprintf(stderr, "Expected body!");
          abort();
        }


        body_received += frame.payload.body_fragment.len;
        count++;
        assert(body_received <= body_target);

//        amqp_dump(frame.payload.body_fragment.bytes,
//                  frame.payload.body_fragment.len);
      }
      printf("fr count:%d\n", count);
      int array[count];
      printf("size of array %d\n",sizeof(array));
      if (body_received != body_target) {
        /* Can only happen when amqp_simple_wait_frame returns <= 0 */
        /* We break here to close the connection */
        break;
      }
    }
  }

  die_on_amqp_error(amqp_channel_close(conn, 1, AMQP_REPLY_SUCCESS), "Closing channel");
  die_on_amqp_error(amqp_connection_close(conn, AMQP_REPLY_SUCCESS), "Closing connection");
  die_on_error(amqp_destroy_connection(conn), "Ending connection");

  return 0;
}
```

**Comment by alanxz** [comment]
This will hopefully be improved when #259 is implemented.

**Comment by alanxz** [comment]
This is probably fixed as of #256 and #260.
